### ISSUE TITLE
Boolean values are not inferred correctly when using the JSON type

### ISSUE THREAD

**Issue by melvynator** [ClickHouse/ClickHouse#37206]
**Describe what's wrong**

Boolean value is inferred as UInt64 instead of boolean

**Does it reproduce on recent release?**

Yes

**How to reproduce**

Let's create a table:

```
CREATE TABLE testJson
(
    `id` UInt32,
    `fields` JSON
)
ENGINE = ReplicatedMergeTree
ORDER BY id
SETTINGS index_granularity = 8192
```

Let's add a row with a boolean value:

```
INSERT INTO testJson  (id, fields) VALUES (1,'{"fieldInteger":1,"fieldBoolean":true,"fieldFloat":1.23,"fieldString":"StringValue"}');

Query id: d43fbeac-fb7f-47c2-b543-bd49b524a745

Ok.

1 rows in set. Elapsed: 0.385 sec.
```

Let's see how the boolean value was inferred:

```
SELECT
    fields.fieldBoolean,
    toTypeName(fields.fieldBoolean)
FROM testJson

Query id: 5aded79a-c7dd-4628-8cc1-5f338e5525fc

┌─fields.fieldBoolean─┬─toTypeName(fields.fieldBoolean)─┐
│                   1 │ UInt64                          │
└─────────────────────┴─────────────────────────────────┘

1 rows in set. Elapsed: 0.204 sec.
```

It has been inferred as: `UInt64`

* Which ClickHouse server version to use
`ClickHouse 22.5.1.4150  `

**Expected behavior**
It should have been inferred as `boolean`

**Comment by CurtizJ** [comment]
I've just copy-pasted your commands and it works as expected:

```sql
CREATE TABLE testJson
(
    `id` UInt32,
    `fields` JSON
)
ENGINE = ReplicatedMergeTree
ORDER BY id
SETTINGS index_granularity = 8192;
                    
INSERT INTO testJson  (id, fields) VALUES (1,'{"fieldInteger":1,"fieldBoolean":true,"fieldFloat":1.23,"fieldString":"StringValue"}');
                   

SELECT
    fields.fieldBoolean,
    toTypeName(fields.fieldBoolean)
FROM testJson;

┌─fields.fieldBoolean─┬─toTypeName(fields.fieldBoolean)─┐
│                   1 │ UInt8                           │
└─────────────────────┴─────────────────────────────────┘

SELECT version();

┌─version()───┐
│ 22.5.1.4150 │
└─────────────┘
```

**Comment by melvynator** [comment]
I've created another service and was still able to recreate my issue:

```
SET allow_experimental_object_type = 1

Query id: b5b677ac-f1d5-44a1-8b8b-c1ccd3cc60f0

Ok.

0 rows in set. Elapsed: 0.157 sec.
```

```
CREATE TABLE testJson
(
    `id` UInt32,
    `fields` JSON
)
ENGINE = ReplicatedMergeTree
ORDER BY id
SETTINGS index_granularity = 8192

Query id: 322cc08c-f500-451f-8fda-755528b68f07

┌─host───────────────────────────┬─status─┬─error─┬─num_hosts_remaining─┬─num_hosts_active─┐
│ default|c-black-ew-68-server-0 │      0 │       │                   1 │                1 │
└────────────────────────────────┴────────┴───────┴─────────────────────┴──────────────────┘
┌─host───────────────────────────┬─status─┬─error─┬─num_hosts_remaining─┬─num_hosts_active─┐
│ default|c-black-ew-68-server-1 │      0 │       │                   0 │                0 │
└────────────────────────────────┴────────┴───────┴─────────────────────┴──────────────────┘

2 rows in set. Elapsed: 0.500 sec.
```

```
INSERT INTO testJson  (id, fields) VALUES (1,'{"fieldInteger":1,"fieldBoolean":true,"fieldFloat":1.23,"fieldString":"StringValue"}');


Query id: c9db102b-3afa-4c5a-bd53-fed9291c5894

Ok.

1 rows in set. Elapsed: 0.375 sec.
```

```
SELECT
    fields.fieldBoolean,
    toTypeName(fields.fieldBoolean)
FROM testJson

Query id: b187045c-c15b-4bd9-bd39-754ccaddb05a

┌─fields.fieldBoolean─┬─toTypeName(fields.fieldBoolean)─┐
│                   1 │ UInt64                          │
└─────────────────────┴─────────────────────────────────┘

1 rows in set. Elapsed: 0.210 sec.
```

```
SELECT version()

Query id: 12e52c7f-e565-495b-90cb-0a371121afaf

┌─version()───┐
│ 22.5.1.4150 │
└─────────────┘

1 rows in set. Elapsed: 0.158 sec.
```

**Comment by CurtizJ** [comment]
Ah, I forgot that parsing and type inference is done on client in this query. Maybe you have too old client or I have too fresh. Anyway, it has already been fixed in master in #36756.
